doc/contributing.rst

```diff
@@ -1,221 +1,760 @@
+`check it out`_ so that any changes you make will exist on that new branch
+(instead of on ``master``):
+
+.. code-block:: console
+
+    $ git checkout master            # switch to master branch
+    $ git fetch upstream             # get the current state of the upstream repo
+    $ git merge upstream/master      # synchronize local master branch with upstream master branch
+    $ git checkout -b new-feature-x  # create branch "new-feature-x" and check it out
+
+.. note::
+
+    You can save some typing by using ``git pull upstream/master`` to replace
+    the ``fetch`` and ``merge`` lines above.
+
+Now that you're on a new branch, you can fix a bug or add a new feature, add a
+test, update the documentation, etc. When you're done, it's time to organize
+your changes into a series of `commits`_. Commits are like snapshots of the
+repository — actually, more like a description of what has to change to get
+from the most recent snapshot to the current snapshot.
+
+Git knows that people often work on multiple changes in multiple files all at
+once, but that ultimately they should separate those changes into sets of
+related changes that are grouped together based on common goals (so that it's
+easier for their colleagues to understand and review the changes). For example,
+you might want to group all the code changes together in one commit, put new
+unit tests in another commit, and changes to the documentation in a third
+commit.  Git makes this easy(ish) with something called the `stage`_ (or
+*staging area*). After you've made some changes to the codebase, you'll have
+what git calls "unstaged changes", which will show up with the `status`_
+command:
+
+.. code-block:: console
+
+    $ # see what state the local copy of the codebase is in
+    $ git status
+
+Those unstaged changes can be `added`_ to the stage one by one, by either
+adding a whole file's worth of changes, or by adding only certain lines
+interactively:
+
+.. code-block:: console
+
+    $ # add a whole file's worth of changes
+    $ # (same command works to add a completely new file):
+    $ git add mne/some_file.py
+    $ # enter interactive staging mode, to add only portions of the file:
+    $ git add -p mne/viz/some_other_file.py
+
+Once you've collected all the related changes together on the stage, the ``git
+status`` command will now refer to them as "changes staged for commit". You can
+commit them to the current branch with the `commit`_ command. If you just type
+``git commit`` by itself, git will open the text editor you configured it to
+use so that you can write a *commit message* — a short description of the
+changes you've grouped together in this commit. You can bypass the text editor
+by passing a commit message on the command line with the ``-m`` flag. For
+example, if your first commit adds a new feature, your commit message might be:
+
+.. code-block:: console
+
+    $ git commit -m 'ENH: adds feature X to the Epochs class'
+
```

Comment:
Since you already elaborate quite a bit on the commit message, one thing I find very useful is `git commit --amend` to edit the last commit message.

Reply:
that opens up a big rabbit hole, because if users have already pushed, `--amend` will require a subsequent `push --force`.  And `--amend` also will incorporate changes staged since the commit, not just change the commit message.  I think discussion of `--amend` should be reserved for a separate "git troubleshooting" section or page, along with `rebase`

Reply:
I see your point, very valid. (I only use it directly after commiting.)